snuba/optimize.py

```diff
@@ -291,18 +341,58 @@ def optimize_partitions(
 
         query = (query_template % args).strip()
         logger.info(f"Optimizing partition: {partition}")
-        start = time.time()
 
         # Update tracker before running clickhouse.execute since its possible
         # that the connection can get disconnected while the server is still
         # executing the optimization. In that case we don't want to
         # run optimization on the same partition twice.
         if tracker:
-            tracker.update_completed_partitions(partition)
+            tracker.update_scheduled_partitions(partition)
 
+        # if theres a merge in progress for this partition, wait for it to finish
+        while True:
+            busy_merging, estimated_sleep_time = is_busy_merging(
+                clickhouse, database, table
+            )
+            if not busy_merging:
+                break
+            else:
+                time.sleep(OPTIMIZE_BASE_SLEEP_TIME + estimated_sleep_time)
+
+        start = time.time()
         clickhouse.execute(query, retryable=False)
         metrics.timing(
             "optimized_part",
             time.time() - start,
             tags=_get_metrics_tags(table, clickhouse_host),
         )
+
+
+def is_busy_merging(
+    clickhouse: ClickhousePool, database: str, table: str
+) -> Tuple[bool, float]:
+    """
+    Returns true and the estimated sleep time if clickhouse is busy with merges in progress
+    for the table. Clickhouse is considered busy if
+        1. there are more than OPTIMIZE_MERGE_MAX_CONCURRENT_JOBS merges in progress
+        2. or there is a merge of size greater than OPTIMIZE_MERGE_SIZE_CUTOFF
+    """
+    merge_info = get_current_merging_partitions_info(clickhouse, database, table)
+
+    if len(merge_info) > OPTIMIZE_MERGE_MAX_CONCURRENT_JOBS:
+        estimated_sleep_time = max(
+            merge_info, key=lambda x: x.estimated_time
+        ).estimated_time
+        logger.info(
+            f"too many concurrent merges {len(merge_info)}, sleeping for {estimated_sleep_time}s"
+        )
+        return True, estimated_sleep_time
```

Comment:
We should not have this check. It is possible for clickhouse to run many small merges concurrently. Those have no impact on the system. We should pause only if big merges are ongoing.

Reply:
I updated `get_current_merging_partitions_info` to return only long running merges as Filippo suggested then sleep if we have too many of those going on. Should I only check size of merges?

Reply:
Ok. I think we can improve readability of the code. `get_current_merging_partitions_info` should be called something like `get_long_merges_info` or something. The current name suggests that it gets all of the current merges.